src/mame/machine/eispc_kb.cpp

```diff
@@ -0,0 +1,446 @@
+//  ROM( eispc_kb )
+//-------------------------------------------------
+
+ROM_START( eispc_kb )
+	ROM_REGION( 0x800, M6801_TAG, 0 )
+	ROM_LOAD( "sgtpepper-1.2.bin", 0x000, 0x800, CRC(7107b841) SHA1(a939dd50622575c31fea9c7adb7a7db5403a7aca) )
+ROM_END
+
+} // anonymous namespace
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(EISPC_KB, eispc_keyboard_device, "eispc_kb", "Ericsson PC keyboard")
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  eispc_keyboard_device - constructor
+//-------------------------------------------------
+
+eispc_keyboard_device::eispc_keyboard_device(
+		machine_config const &mconfig,
+		char const *tag,
+		device_t *owner,
+		uint32_t clock)
+	: device_t(mconfig, EISPC_KB, tag, owner, clock)
+	, m_mcu(*this, M6801_TAG)
+	, m_rows(*this, "P1%u", 0)
+	, m_txd_cb(*this)
+	, m_rxd_high(true)
+	, m_txd_high(true)
+	, m_hold(true)
+	, m_col_select(0)
+{
+}
+
+INPUT_CHANGED_MEMBER( eispc_keyboard_device::key )
+{
+		if (oldval && !newval)
+	{
+		LOGUI("Key Pressed - name: %s field: %04x param: %04x oldval: %04x newval: %04x\n", field.name(), field.defvalue(), param, oldval, newval);
+		int idx = *((int *)(&param));
+		if (idx >= sizeof(keys))
+			logerror("Out of bounds access in keys array\n");
+		else
+			keys[idx] |= (uint16_t) field.defvalue();
+	}
+	else if (newval && !oldval)
+	{
+		LOGUI("Key Released - name: %s field: %04x param: %04x oldval: %04x newval: %04x\n", field.name(), field.defvalue(), param, oldval, newval);
+		int idx = *((int *)(&param));
+		if (idx >= sizeof(keys))
+			logerror("Out of bounds access in keys array\n");
+		else
+			keys[idx] &= ~(uint16_t)field.defvalue();
+	}
+	for (int i = 0; i < 6; i++) LOGUI("%04x ", keys[i]); LOGUI("\n");
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::rxd_w)
+{
+	LOGBITS("KBD bit presented: %d\n", state);
+	m_rxd_high = CLEAR_LINE != state;
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::hold_w)
+{
+	m_hold = CLEAR_LINE == state;
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::rst_line_w)
+{
+	if (state == CLEAR_LINE)
+	{
+		m_mcu->resume(SUSPEND_REASON_RESET);
+		LOGRST("KBD: Keyboard mcu reset line is cleared\n");
+	}
+	else
+	{
+		m_mcu->suspend(SUSPEND_REASON_RESET, 0);
+		LOGRST("KBD: Keyboard mcu reset line is asserted\n");
+	}
+
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void eispc_keyboard_device::device_reset()
+{
+	LOGRST("KBD: Keyboard is in reset until host computer explicitly releases the reset line\n");
+	m_mcu->suspend(SUSPEND_REASON_RESET, 0);
+	for (auto & elem : keys) elem = 0;
```

Comment:
The state of inputs doesn't change when the device is reset.

Reply:
ok, but I want to keep the MCU in reset until released by the host computer, so I have to implement a reset line for the 6801 then? But even so, the line must be kept there from reset until it is released, hmmm. The MCU may  be reset and kept there at any time really.

Reply:
AFAIK the reset line is implemented in the MAME core, not in individual CPUs.  If you assert it MAME stops scheduling the CPU to run, and then when you release it the CPU's reset handler is called and it starts running again.

Reply:
ok, explains why I couldn't find it then, I'll have a look

Reply:
Ok, so this is what I found:

I can remove the suspend() in the device_reset as mentioned above but if I change the suspend()/resume() calls in rst_line_w() to set_input_line() calls I get Keyboard error. This also happens if I change the argument in suspend() to true instead of 0/false which affects the eatcycles behavior. 

I noticed that the keyboard and the host spontaneously sends one byte each to each other at startup, at close to the same time and check that each side received it, so this can be a timing issue which I accidentally avoided by using suspend/resume instead of set_input_line (which basically does the same thing but with some more glue and setting the eatcycles to true)

So I am inclined to commit the removal of the reset code but keep the susend()/resume()

Reply:
The scheduler bites us again!  There are too many drivers where we have to work around the scheduler.

Reply:
I wonder if the same scheduling issue could also be the cause of the unexpected watchdog reset in Elevator Action (MT 7399) and the ACIA race condition that I had to work around in balsente.cpp.

Reply:
With the input state though, you're keeping a copy of the current state of the key inputs, and you're clearing that on reset.  This is not going to work properly.  If a key is held down across reset, it's held down across reset.  You're not going to magically get an input changed call when you take the MCU out of reset to remind you that the key is held down.

There are two ways to deal with MCU input ports: pushed and pulled.  For the pushed approach, when anything happens that would affect the value the MCU sees, you compute the new value and push it to the MCU.  For the pulled approach, you wait for the MCU to tell you it's reading the input port and then compute the value it would see.

You've got a half-and-half setup in this driver - you've got `m_keys` that you update when an input changes, but you wait for `in_p1_cb` to give a value to the MCU.  You should probably just go for the pulled approach and read the input ports in your `in_p1_cb` rather than using `m_keys`.  I guess you can keep the input changed handler around for logging if you really want, but as it's being used now, it's making the code a lot more complex and error-prone.

Reply:
Thx for the input, I saw some drivers using timers to tweak the timing of the initial handshake, maybe I can do that instead of touching the scheduler. I just found that the alternative firmware for epc still has the Keyboard error so I need to track that down anyway. 

Yes I can definitely remove the m_keys now when I know how the keyboard scan works and call read() on p1_r instead, it was just helpful for a while.